content/blog/2023-12-11-jitnic/index.md

```diff
@@ -0,0 +1,129 @@
+
+This trace and path condition will get installed onto the NIC. Now, if a packet arrives at the NIC and it matches the path condition, it will simply run the straight line trace. A simple DCE pass could remove any instructions that are only used in conditions. More complex programs might also do some packet writes in the trace, producing longer traces than just a forwarding decision. 
+
+Finally, if an ipv6 packet arrives, it will not match the path condition and need to be sent to the host CPU to be compiled. After tracing, the CPU will install a new trace and path condition on the smartNIC. At this point, ipv6 and ipv4 packets will match their respective path conditions and can be run on the NIC without going to the compiler.
+
+## Evaluation
+
+Considering our smartNIC is just a software simulator, it does not make sense to evaluate the runtime performance of our system. Instead, we focus on the “cache” performance we can obtain. In real hardware the size and replacement policy of the cache could vary and have important ramifications for system performance – every cache miss requires the compiler to run and install new rules into the smartNIC. 
+
+The cache performance depends on the path conditions and the sequence of packets arriving at the NIC. If a large number of packets arrive that all match the same few path conditions, we will have a great cache hit rate. There is a worst case scenario where the size of the cache is less than the number of possible traces of a program and a bad sequence of packets causes misses to continuously occur. We call this behavior “cache thrashing”. In the ipv6 filter case study, considering only well formed packets, we only need 2 traces to handle the entire state space of input packets. However, if a badly formed, non-ethernet, packet arrives, the program will fail the bounds check and need to be traced by the compiler. Assuming the trace cache on the NIC has 2 entries, this new trace will evict one of the previous entries. Without loss of generality, assume it is the ipv6 case. Now, if an ipv6 packet arrives the program will need to be re-traced, which will evict another trace, etc. It is clear how an “unlucky” ordering of packets can lead to continuous misses. 
+
+This emphasizes the importance of tuning the cache size - which could potentially dynamically change depending on the program. It also shows that the replacement policy can potentially affect the probability of experiencing cache thrashing. In practice, we find XDP programs to be quite short and not have many conditions, meaning a small trace cache can fit most possible traces. 
```

Comment:
> In practice, we find XDP programs to be quite short and not have many conditions, meaning a small trace cache can fit most possible traces.

Can you say a little bit more about what evidence you gathered for this assertion? How many programs did you try, and where did you find them?

Reply:
Good point, we should have included this from the start. I've updated the post. In summary, we don't currently support BPF maps, so we couldn't run many programs outside of stateless filtering ones or contrived examples. To reach these conclusions, we manually inspected XDP programs that are included in the Linux kernel, instead of running them in our system. I made this clear and discussed the downsides.